### PR TITLE
Spec for authentication experience for VS Online Environments

### LINKED ISSUE #9244: Design the authentication experience for VS Online Environments
Currently the authentication model is talk to a plugin/pop-up a dialog if needed. 
We need to ensure that the customers receives a dialog similarly to normal VS. 
[1073273](https://devdiv.visualstudio.com/DevDiv/_queries/edit/1073273/)

This should be converted to an epic if necessary.

### PR DESCRIPTION
**Pull request by rrelyea**
To Fix: https://github.com/nuget/home/issues/9244

### CODE REVIEW

File: designs/Remote-Authentication.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ 1. VS Auth - Inside VS, NuGet can demand implementers of Auth. Azure Artifacts ships one with VS.
+ 1. NuGet CLI v1 Auth - NuGet.exe historically allowed an exe in the same directory that nuget.exe would communicate with.
+ 1. NuGet CLI v2 Auth - In NuGet 4.8 (VS 15.8), we enabled a new xplat auth plugin technique. This allows xplat auth in all of our codepaths (nuget.exe, msbuild.exe, dotnet.exe and VS).
+
+ When solving this problem when there is some nuget code running on the client and server, we need to decide which of those to support.
+
+## Possible approaches
+
+1. Remote v2 Auth call to client - When credentials are needed, the server calls to the client to authenticate via prompts, etc... Relies on NuGet CLI v2 Auth provider being installed on VS client side.
+2. support VS auth providers too? - not sure if this parallel code base is important to support.
```
Comment (nkolev92) [inline designs/Remote-Authentication.md]: I think it's up to us. 
We'll need a server side coalescing NuGet provider and given that VS supports V2 providers too, we can do it with only 1 NuGet provider if we wanted to.

File: designs/Remote-Authentication.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+ 1. VS Auth - Inside VS, NuGet can demand implementers of Auth. Azure Artifacts ships one with VS.
+ 1. NuGet CLI v1 Auth - NuGet.exe historically allowed an exe in the same directory that nuget.exe would communicate with.
+ 1. NuGet CLI v2 Auth - In NuGet 4.8 (VS 15.8), we enabled a new xplat auth plugin technique. This allows xplat auth in all of our codepaths (nuget.exe, msbuild.exe, dotnet.exe and VS).
+
+ When solving this problem when there is some nuget code running on the client and server, we need to decide which of those to support.
+
+## Possible approaches
+
+1. Remote v2 Auth call to client - When credentials are needed, the server calls to the client to authenticate via prompts, etc... Relies on NuGet CLI v2 Auth provider being installed on VS client side.
```
Comment (nkolev92) [inline designs/Remote-Authentication.md]: I don't know if we need to do this, or even if we can to an extent :) 

V2 providers + msbuild.exe & dotnet.exe rely on device flow, and the logs are automatically visible, so there's no remoting necessary.

**Review by nkolev92 — COMMENTED** [review]
Ping me to review again as you add more content in here.

File: designs/Remote-Authentication.md (hunk 2 above)
Comment (jmyersmsft) [reply designs/Remote-Authentication.md]: Device flow is a really ugly experience and imo looks like the devs are incompetent when it appears in scenarios that are actually interactive.

Comment (jmyersmsft) [reply designs/Remote-Authentication.md]: especially if this is supposed to be "seamless", Device Flow is a really big seam.

Comment (jmyersmsft) [reply designs/Remote-Authentication.md]: If I could make a clone of myself to just work on artifacts-credprovider I would do everything I possibly could to present the user with _anything_ other than device flow. When we eventually move from ADAL to MSAL for authenticating to AAD, I believe there are more options for interactive auth UX even in x-plat netcore land there.

**Comment by rrelyea** [comment]
Added a few scenarios to the spec.

**Comment by Penguinwizzard** [comment]
To copy over a note from an email chain elsewhere - it'd be interesting to see authentication to nuget feeds in containerized scenarios (where the host may have auth information, but needs to provide a way for the containerized build process to access artifacts from the authenticated feeds). This additionally can be broken down into VS scenarios (where ideally the process would look the same on a build-per-build basis to the end user in a containerized scenario as in a non-containerized scenario - using the VS auth mechanisms) and nuget cli scenarios (for all non-VS build scenarios, where it'd ideally do something magic that I can't think of right now). Might not be in-scope for this document, though.

**Comment by JonDouglas** [comment]
Hello!

I'm working to cleanup stale PRs on `NuGet/Home`. 

Actions to take:

- If this PR **should be merged** into `NuGet/Home`, please assign a handful of reviewers or ask for another review to approve a merge.
- If this PR **should be abandoned**, please close it.
- If this PR **should remain open**, please bump the conversation forward or leave a comment for this to not be considered abandoned in a future pass.

If there's anything I can do to help out, please let me know!

**Comment by JonDouglas** [comment]
I am closing this PR due to staleness. If this PR should be merged into the repo, please re-open & assign appropriate reviewers. Thanks!

**Pull request closed without merging**